Userland/Libraries/LibThreading/BackgroundAction.cpp

```diff
@@ -15,27 +15,29 @@ static pthread_mutex_t s_mutex = PTHREAD_MUTEX_INITIALIZER;
 static pthread_cond_t s_condition = PTHREAD_COND_INITIALIZER;
 static Queue<Function<void()>>* s_all_actions;
 static Threading::Thread* s_background_thread;
+static Atomic<bool> s_background_thread_should_run = true;
 
 static intptr_t background_thread_func()
 {
     Vector<Function<void()>> actions;
-    while (true) {
-
+    while (s_background_thread_should_run.load(AK::MemoryOrder::memory_order_acquire)) {
         pthread_mutex_lock(&s_mutex);
 
-        while (s_all_actions->is_empty())
+        while (s_all_actions->is_empty() && s_background_thread_should_run.load(AK::MemoryOrder::memory_order_acquire))
```

Comment:
@bugaevc is there an easier way to make this background thread interruptible? When I started using it in ImageDecoder, Ladybird wouldn't exit until all threads of ImageDecoder (its child) were exited.

Reply:
Well, there's `pthread_cancel()`, if you make image decoding check for cancellation regularly.

Reply:
Ehhh. Everything I've read about pthread_cancel suggests that it plays very poorly with C++ code. iirc libstdc++ (used to?) even throw a private exception type to get unwinding on cancellation. Adding tests for pthread_testcancel() or setting up pthread_setcancelstate() all over LibGfx seems way less palatable than a few extra atomic reads.

Reply:
> iirc libstdc++ (used to?) even throw a private exception type to get unwinding on cancellation.

Yes, that's how cancellation is implemented in NPTL (which is not libstdc++) -- it calls `_Unwind_ForcedUnwind()`, which is seen as an exception from C++. This also handles C code that does `pthread_cleanup_push()`, objc `@try` / `@catch`, runs Rust destructors, etc. We don't have that in HTL though 😢

> Adding tests for pthread_testcancel() or setting up pthread_setcancelstate() all over LibGfx seems way less palatable than a few extra atomic reads.

Sure, but does your thread not block inside LibGfx for a long time? If it does, are you going to run these atomic checks in LibThreading at all?

Reply:
It blocks for on the order of hundreds of milliseconds for each job. The important time horizon is application shutdown, where taking ~1sec is not a big deal, as long as we eventually do shutdown the thread and clean up the application properly.